core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java

```diff
@@ -387,4 +388,163 @@ public void dataFilesCleanup() throws IOException {
     Assert.assertTrue("FILE_A should be deleted", deletedFiles.contains(FILE_A.path().toString()));
     Assert.assertTrue("FILE_B should be deleted", deletedFiles.contains(FILE_B.path().toString()));
   }
+
+  /**
+   * Test on table below, and expiring the staged commit `B` using `expireOlderThan` API.
+   * Table: A - C
+   *          ` B (staged)
+   */
+  @Test
+  public void testWithExpiringDanglingStageCommit() {
+    // `A` commit
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    // `B` staged commit
+    table.newAppend()
+        .appendFile(FILE_B)
+        .stageOnly()
+        .commit();
+
+    TableMetadata base = readMetadata();
+    Snapshot snapshotB = base.snapshots().get(1);
+
+    // `C` commit
+    table.newAppend()
+        .appendFile(FILE_C)
+        .commit();
+
+    List<String> deletedFiles = new ArrayList<>();
+
+    // Expire all commits including dangling staged snapshot.
+    table.expireSnapshots()
+        .deleteWith(deletedFiles::add)
+        .expireOlderThan(snapshotB.timestampMillis() + 1)
+        .commit();
+
+    // Files should be deleted of dangling staged snapshot
+    snapshotB.addedFiles().forEach(item -> {
+      Assert.assertTrue(deletedFiles.contains(item.path().toString()));
+    });
+  }
+
+  /**
+   * Expire cherry-pick the commit as shown below, when `B` is in table's current state
+   *  Table:
+   *  A - B
+   *   `- C (B) <--this commit
+   */
+  @Test
+  public void testWithCherryPickTableSnapshot() {
+    // `A` commit
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    // `B` commit
+    table.newAppend()
+        .appendFile(FILE_B)
+        .commit();
+    // pick the snapshot 'B`
+    TableMetadata base = readMetadata();
+    Snapshot snapshotA = base.snapshots().get(0);
+    Snapshot snapshotB = base.snapshots().get(1);
+
+    // Move the table back to `A`
+    table.manageSnapshots()
+        .setCurrentSnapshot(snapshotA.snapshotId())
+        .commit();
+
+    // Generate A -> `C (B)`
+    table.manageSnapshots()
+        .cherrypick(snapshotB.snapshotId())
+        .commit();
+    Snapshot snapshotC = readMetadata().currentSnapshot();
+
+    // Move the table back to `B`
+    table.manageSnapshots()
+        .setCurrentSnapshot(snapshotB.snapshotId())
+        .commit();
+    List<String> deletedFiles = new ArrayList<>();
+
+    // Expire `C`
+    table.expireSnapshots()
+        .deleteWith(deletedFiles::add)
+        .expireSnapshotId(snapshotC.snapshotId())
```

Comment:
You can simplify this and avoid using `expireSnapshotId` by using `expireOlderThan` a future time (`Long.MAX_VALUE`). Snapshot B won't expire because it is the current snapshot.